src/StoreMeta.h

```diff
@@ -80,67 +64,96 @@ enum {
         uint16_t flags;
      \endcode
      */
-    STORE_META_STD,
-
-    /**
-     * Reserved for future hit-metering (RFC 2227) stuff
-     */
-    STORE_META_HITMETERING,
-
-    // TODO: document this TLV type code
-    STORE_META_VALID,
+    STORE_META_STD = 5,
 
     /**
      * Stores Vary request headers
      */
-    STORE_META_VARY_HEADERS,
+    STORE_META_VARY_HEADERS = 8,
 
     /**
      * Updated version of STORE_META_STD, with support for  >2GB objects.
      * As STORE_META_STD except that the swap_file_sz is a 64-bit integer instead of 32-bit.
      */
-    STORE_META_STD_LFS,
+    STORE_META_STD_LFS = 9,
 
     // TODO: document this TLV type code
-    STORE_META_OBJSIZE,
+    STORE_META_OBJSIZE = 10
 
-    STORE_META_STOREURL,    /**< the Store-ID url, if different to the normal URL */
-    STORE_META_VARY_ID,     /**< Unique ID linking variants */
-    STORE_META_END
+    // When adding values, update Store::SwapMetaTypeMax.
+    // When removing values, check Store::IgnoredSwapMetaType() and friends.
```

Comment:
Using the enum labels properly instead of magic numbers everywhere will fix this maintenance problem as well as avoid future bugs.

Reply:
Keeping deprecated/unused enum values in the enum makes it impossible to write safe type-specific code without a bunch of ugly `assert(false)` statements in _every_ `switch`. We have been there before this PR. This PR makes things _better_ by hiding deprecated values from the rest of the code and allowing the compiler to check some of the `switch` statements.

Asking developers to move a deprecated name from one place to another is not really a "maintenance problem" (worth discussing) -- the effort is trivial and is essentially _required_ to properly classify the removal anyway.

As for bugs, I think we are well protected here (and much better than the official code is!), especially after my branch commit c39ac63:

* If somebody forgets to remove a now-deprecated value from the enum, the compiler will probably warn (because all `switch` statements ought to drop the _deprecated_ value handling, causing a compiler warning).
* If somebody forgets to add a now-deprecated value to one of the IgnoredSwapMetaType() functions, then "make check" will now fail because/and Squid will assert on startup. I think it will be possible to convert that check (added in c39ac63) into a compile-time static_assert after C++17 migration.

BTW, a documented integer constant that is used only once can be a good idea. It is not a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants) anti-pattern. It is not "magic" because it is documented/explained, and it is not causing code duplication problems common to that anti-pattern because it is used only once. It is just a tool that has its valid uses. As the bug fixed in branch commit ee99367 illustrates, using certain names twice instead of single-use unnamed constants can create code duplication problems!

Reply:
> this comment can be removed

Done in 01ef750.

----

> thanks for the verbose re-wording of what I stated

If that was sarcasm, then you are welcome.

If you were serious, then, FWIW, my verbose comment disagrees with (and proves the opposite of) what you stated.

> Now that the other code is using these labels instead of numbers ...

The "other code" is not using *these* labels/values and have never used them! The three sets of labels/values are disjoint and will always be disjoint, by PR design: honored, deprecated, reserved. We have assertions that ensure that principle is upheld. Whether that "other code" is using numbers, labels, external function results, or voodoo has absolutely no bearing on any of this!

If "these" in "using these labels" was a typo and should be dropped/ignored, then I do not understand the connection between the (different) labels used in that "other code" and the now-removed comment. That "other code" is not using any labels present in _this_ enum (and will never use any labels present in any contemporary SwapMetaType enum).

Reply:
> > this comment can be removed
> 
> Done in [01ef750](https://github.com/squid-cache/squid/commit/01ef750e7950dcb20141927ffb622fa20d716359).
> 
> > thanks for the verbose re-wording of what I stated
> 
> If that was sarcasm, then you are welcome.
> 
> If you were serious, then, FWIW, my verbose comment disagrees with (and proves the opposite of) what you stated.
> 

I was serious, perhapse you misunderstood the request.


> > Now that the other code is using these labels instead of numbers ...
> 
> The "other code" is not using _these_ labels/values and have never used them! The three sets of labels/values are disjoint and will always be disjoint, by PR design: honored, deprecated, reserved.

Indeed. And now that they are labelled as such we do not have to remember exactly what integer values are being used by the honored enum and which are not when reviewing if statements around the code.

Defining the deprecated and reserved enums right next to the larger if statements using their values makes the code look odd. BUT, those definitions technically should be next to the honored enum definition which would make the magic-value situation more obvious.